### PR TITLE
feat: core side of protocol upgrade

### LINKED ISSUE #5538: Core side of protocol upgrade
# Feature Overview  (part of #5004)

- In Admin RPC
   - Add API for indicating interest to upgrade the protocol with a release tag and a upgrade block height: `ProposeNewUpgrade(upgradeBlockHeight uint64, releaseTag string) error`. In response the core submits a transaction to vega chain so that all nodes see the upgrade request. The `upgradeBlockHeight` is the last block to be processed in the chain before the chain will go down in case it is approved. 
   - Add API to query if the protocol is ready to be shutdown for upgrade: `IsReadyForUpgrade() bool`
   
- An engine is added to keep track of protocol upgrade proposals
- If a proposal doesn't get a majority of +2/3 of the *tendermint* validators until (including) the block height given in `upgradeBlockHeight` of the proposal - the proposal is rejected and dropped. 
- If it gets a majority of +2/3 of the *tendermint* validators - at the beginning of the following block - we take a snapshot at the end of which `IsReadyForUpgrade` RPC call will start returning true.

### PR DESCRIPTION
**Pull request by ze97286**
Closes #5538

### CODE REVIEW

File: protocolupgrade/engine.go
Comment (jeremyletang) [inline protocolupgrade/engine.go:35]: Shouldn't we use the network parameters instead for this?

File: processor/abci.go
Comment (wwestgarth) [inline processor/abci.go:620]: I don't think we can take the snapshot here. Assuming I follow properly we are at block `n`, the first block that will be in the new version, and effectively snapshoting block`n-1`, the last block of the old version.

When we pull down the old core process and bring up the new one with the snapshot data from block `n-1` tendermint will want the commit hash of block `n-1` during the abci `Info` call, which we won't have.  So I think you'll have to move the forcing of the snapshot to `OnCommit` of block `n-1`.

**Comment by wwestgarth** [comment]
Not for this PR, but I think we are going to have to be careful more careful with which snapshots are sent/accepted when state-syncing. For example if someone if running an old version of core and state-syncs against a network with a newer version it'll just get handed a snapshot and it will just try to use it. We'll want to put something in to reject any snapshots from different versions.

It might also make sense to force a snapshot on the first block after an upgrade otherwise any one wanting to state-sync in will have to wait until the first snapshot is taken before getting anything usable.

File: processor/abci.go
Comment (wwestgarth) [inline processor/abci.go:912]: I'm glad to see the update height is not longer compiled in now, but who is now in charge of choosing it?

Comment (jeremyletang) [reply processor/abci.go:912]: Validators should discuss it together and fix a block height together.

**Comment by wwestgarth** [comment]
> > Not for this PR, but I think we are going to have to be careful more careful with which snapshots are sent/accepted when state-syncing. For example if someone if running an old version of core and state-syncs against a network with a newer version it'll just get handed a snapshot and it will just try to use it. We'll want to put something in to reject any snapshots from different versions.
> > It might also make sense to force a snapshot on the first block after an upgrade otherwise any one wanting to state-sync in will have to wait until the first snapshot is taken before getting anything usable.
> 
> I think this should be covered with this? https://github.com/vegaprotocol/vega/blob/develop/processor/abci.go#L391

Maybe, but the info call happens after we would have already loaded the snapshot. 

That is a good point though, we probably need to set the new protocol-version in the consensus-params in the EndBlock response on the block before we upgrade.

File: protocolupgrade/engine.go
Comment (ze97286) [reply protocolupgrade/engine.go:35]: also, I think each governance proposal type has its own so we can add a new one

File: admin/server.go
Comment (karlem) [inline admin/server.go:103]: I don't think this is enough. I think you will need to implement a proper service like the one we use for `nodeWallet`.  With http request, args and reply.

Comment (wwestgarth) [reply admin/server.go:103]: Is there a reason we aren't just going with a CLI thing like `vega propose_upgrade --release-tag=blah --block=blah` to shoot the command in like when a new node annouces itself?

File: protocolupgrade/engine.go
Comment (karlem) [inline protocolupgrade/engine.go:105]: Not sure what that actually means but would that not be an issue for non-validators (maybe those who runs data-node)?

File: admin/server.go
Comment (ze97286) [reply admin/server.go:103]: we may, none of this is set in stone. whatever ends up simpler for the validators / process manager

File: protocolupgrade/engine.go
Comment (ze97286) [reply protocolupgrade/engine.go:186]: non-validators don't decide if the network is upgraded so they can't "vote" or propose an upgrade. They will be running a process manager so should react the same way as validators. The problem is that if we're not automating the binaries preparation then they need to know that they need to have built the binaries by this point and then the process manager will restart correctly.

File: protocolupgrade/engine.go
Comment (ze97286) [reply protocolupgrade/engine.go:35]: ok, I'll add a network parameter for it.

File: protocolupgrade/snapshot.go
Comment (ze97286) [reply protocolupgrade/snapshot.go:38]: I'm not sure what's the best strategy is. Maybe for simplicity - once a proposal gets approved - we're locking the state, i.e. rejecting everything else, so the state of the engine is clear, and auto-rejecting any incoming proposal until the upgrade takes place. what do you think? @wwestgarth @jeremyletang @karlem

Comment (wwestgarth) [reply protocolupgrade/snapshot.go:38]: A fresh start after an upgrade seems reasonable to me.

Comment (ze97286) [reply protocolupgrade/snapshot.go:38]: done

File: admin/server.go
Comment (ze97286) [reply admin/server.go]: done

File: processor/abci.go
Comment (wwestgarth) [inline processor/abci.go:719]: This snapshot hash needs to be set as `resp.Data`. Also if it happens to already be a snapshot block, we'll be double-snapshotting and will have two snapshots for the same block height in the snapshotdb.

Comment (ze97286) [reply processor/abci.go:719]: re double snapshots - I'm not sure how the snapshotdb will handle that, but the reason I force a snapshot here regardless if one was taken or not is because of the asynchronous save of the snapshot which I can't rely on here.

Regarding to setting the snapshot hash in resp.Data, why do we need this? I don't want to force using it for consensus, I just want to make sure it's saved.

Comment (ze97286) [reply processor/abci.go]: ok

File: admin/server.go
Comment (jeremyletang) [inline admin/server.go:119]: I think we have a IsValidator call we could use instead? not a big deal, but the nodewallet would be instatiated evern for a non-validators isn't it?

File: cmd/vega/node/node.go
Comment (jeremyletang) [inline cmd/vega/node/node.go:294]: Alright you do have 2 constructors to instantiate it in a validator or non-validators mode indeed.

File: cmd/vega/propose_upgrade.go
Comment (jeremyletang) [inline cmd/vega/propose_upgrade.go:65]: Should we add validation of the semver here / check if the block is non-zero but also is a future block? Just so we can add nice error handling here.

File: cmd/vega/propose_upgrade.go
Comment (jeremyletang) [inline cmd/vega/propose_upgrade.go:100]: as I suppose we would have the error here to be fair.

File: netparams/defaults.go
Comment (jeremyletang) [inline netparams/defaults.go:32]: Why not just reuse the `ValidatorsVoteRequired` ?

File: processor/abci.go
Comment (jeremyletang) [inline processor/abci.go:685]: What if err != nil ? Is that possible? Should we panic if that was to happen?

Comment (jeremyletang) [reply processor/abci.go:685]: ah it is checked just later...

File: protocolupgrade/engine.go
Comment (jeremyletang) [inline protocolupgrade/engine.go:122]: Not sure we should stop here if we are not a validator ourselves right? Even non-validators needs to process the rest of the state after that?

File: protocolupgrade/engine.go
Comment (jeremyletang) [inline protocolupgrade/engine.go:154]: If a validator registered intent for another release before, should we remove them from the approvers of the release if they register for a new one?

File: protocolupgrade/snapshot.go
Comment (jeremyletang) [inline protocolupgrade/snapshot.go:53]: why not just sort over the protocol IDs ids instead? I means your way is probably better as requires only 1 iteration though

File: cmd/vega/propose_upgrade.go
Comment (ze97286) [reply cmd/vega/propose_upgrade.go:65]: added

File: netparams/defaults.go
Comment (ze97286) [reply netparams/defaults.go:32]: we probably can, but not sure we have to bind it to protocol implementation network param. Anyways, can change to us it and spare one network parameter

File: protocolupgrade/engine.go
Comment (ze97286) [reply protocolupgrade/engine.go:122]: removed

File: protocolupgrade/engine.go
Comment (ze97286) [reply protocolupgrade/engine.go:154]: we can yeah

**Review by jeremyletang — APPROVED** [review]

**Pull request merged**
